### ISSUE TITLE
48-bit compiler generates wrong code for comparison with NULL

### ISSUE THREAD

**Issue by rdos314** [open-watcom/open-watcom-v2#988]
Example code:

```
struct dso_meth_st {
    const char *name;
    int (*pathbyaddr) (void *addr, char *path, int sz);
    void *(*globallookup) (const char *symname);
};

typedef struct dso_meth_st DSO_METHOD;

static DSO_METHOD dso_meth_null = {
    "NULL shared library method"
};

void main()
{
    DSO_METHOD *meth = &dso_meth_null;
    if (meth->pathbyaddr == NULL) {
        return -1;
    }
    return (*meth->pathbyaddr) (0, "", 0);
}
```
The code for the comparision looks like this:

  mov eax,cs
  mov es,[ebp-4]
  mov edx,[ebp-8]
  mov edx,es:[edx+6]
  test edx,edx
  jne fail
  test ax,ax
  jz ok

The first test for edx is successful since edx is 0, but when testing ax the compiler checks if the code selector (cs) is zero, which is never the case.

I built the source for the 32-bit RDOS device driver target, which use a 32-bit compact memory model. The same problem should exist  for 48-bit OS/2 applications too.

**Comment by jmalak** [comment]
I don't thing it is CG bug. Pointer arithmetic and assignment is processed by front-end that it is rather bug in FE.

**Comment by rdos314** [comment]
I tested this code both with the current build and a build from 2020, and the result is the same. I think this is a problem that goes back to 1.9 (possibly longer) and is not something new in the V2 fork.

**Comment by jmalak** [comment]
It is possible that it is old bug.
But I did some change in pointer processing with regards to default code segment that it could be also regression.
I will look on the pointer processing in C compiler.
Did you check it with C++ compiler?

**Comment by rdos314** [comment]
It doesn't work to build 32-bit device-drivers using C++ code, so I cannot check it.

**Comment by jmalak** [comment]
I did a little investigation and it is bug in CG and also in FE.
It is a problem with partial lost of kind of pointer during processing and do far pointer comparison instead of near pointers one.

**Comment by rdos314** [comment]
I think that makes sense. When I tried to reproduce it with basic types it worked as expected. It's also not a problem in other drivers where I have removed a lot of the portability layers. However, I cannot do this with OpenSSL as I need to be able to sync it with future versions to keep up with security updates.

**Comment by jmalak** [comment]
It is mainly problem that for 32-bit targets all was simplified for only near pointers because only flat and small memory models were supported. 32-bit compact model is only one using far pointers for 32-bit and not all these simplifications and mismatches were fixed.

**Comment by rdos314** [comment]
I tested the change, but it doesn't seem to make any difference. I also note that when the code tries to execute the function pointer, it uses a near call. It thus seems like the structure is based on near function pointers rather than far function pointers. Which becomes wrong when the code that checks for NULL is based on far pointers (or rather the extension of the near pointer with cs). The pointer that is linked to a name is a far pointer though. 

Generally, it's invalid for the compiler to convert a near pointer to a far pointer other than when an address of something in the code or dgroup segment is converted. That's because 0 could be a legitimate offset within the code or dgroup segment, while in the flat memory model 0, is always NULL. For far pointers, it should be the segment part being zero that identifies a pointer as NULL.

In short, the 32-bit compact memory model should always allocate space for pointers as 48-bit (far), and never as near.

**Comment by jmalak** [comment]
It is only FE partial fix for pointers storage.
Please wait until it will fix fully.
I will note you.
The problem is that in comparison is compared code pointer (32-bit) with data pointer (48-bit).
(void *)0 is recognized by FE as 48-bit data pointer and it extend code pointer to 48-bit by code segment and compare it
relaxation of NULL pointer is working on 16-bit code but not on 32-bit code.
generally code and data pointers are not compatible but for NULL pointer it should be relaxed

**Comment by rdos314** [comment]
It turns out that other RDOS drivers are affected by this bug too.  It's hidden because the assembler component of those drivers have a function that the linker places first that does a return, and when segment order is fixed, these drivers will fault on null function calls.

**Comment by rdos314** [comment]
Any chance of getting this fixed? I'm not familiar with the code generator so cannot fix it myself. I need to decide if I can have SSL as a device-driver or need to link it to the application. Having SSL as a device-driver requires pointer validations to work properly.

**Comment by jmalak** [comment]
I am now working on code generator to fix as much as possible that this issue is on the list.
Now I start to learn a litle CG code with regards issue #515, #334, #496.
The issue with far pointers will need more work because CG for 32-bit on more places is simplified and works only with offset even if it is far pointer. I already little checked and it is not comparision problem only.

**Comment by rdos314** [comment]
The problem seems to be comparisons with NULL, and not comparisons with 0. When the constant NULL is used the pointer is extended to 48-bits by loading CS, while when comparing with 0, only the 32-bit offset is used. That explains a bit why it's third-party drivers that fail, and not my own since I don't use NULL.

Simplified test code:
`
static void Test1()
{
}

void main()
{
   void (*TestFunc)();

   TestFunc = 0;

   if (TestFunc)
       (*TestFunc)();

   TestFunc = &Test1;

   if (TestFunc)
       (*TestFunc)();

   TestFunc = 0;

   if (TestFunc != NULL)
       (*TestFunc)();
}
`

In the above code, it only fails with the last comparision with NULL which executes the method even if it is set to 0.

The assembly code for if (TestFunc) looks like this:

`
cmp dword ptr [ebp-4],0
jne out
`
The assembly code for if (TestFunc != NULL) looks like this:

`
mov eax,cs
mov edx, [ebp-4]
test edx,edx
jne do
test ax,ax
jne do
jmp out
`

**Comment by jmalak** [comment]
Yes, it could be a workaround. I fixed C compiler in the past that for arithmetic expresion and pointer comparission, pointer is convert to arithmetic value and comparision is arithmetic.
But this far pointer extension doesn't work as should be, somewhere yes and somewhere not.

**Comment by rdos314** [comment]
I think another possible solution is to let the pointer conversion load zero in the segment register instead of CS.  That should work since there should be no far calls anyway, and so the segment part is never used. Still, I don't know how to fix that either. 

I think I will try the work-around to set NULL to 0 for RDOS device drivers, revert back so any calls to CS:0 will result in hitting a breakpoint, and check so there are not more problems related to this.

**Comment by jmalak** [comment]
I reopen it to hold trace of CG bug, even if you use workaround.
